src/java/htsjdk/samtools/SRAFileReader.java

```diff
@@ -0,0 +1,275 @@
+    private static final Log log = Log.getInstance(SRAFileReader.class);
+    private SRAAccession acc;
+    private SAMFileHeader virtualHeader;
+    private ReadCollection run;
+    private ValidationStringency validationStringency;
+    private SRAIterator.RecordRangleInfo recordRangeInfo;
+    private SRAIndex index;
+    private final List<Reference> cachedReferences;
+
+    public SRAFileReader(final File sraFile) {
+        this(new SRAAccession(sraFile.getPath()));
+    }
+
+    public SRAFileReader(final SRAAccession acc) {
+        this.acc = acc;
+
+        if (!acc.isValid()) {
+            throw new RuntimeException("Passes invalid SRA accession into SRA reader: " + acc);
+        }
+
+        try {
+            run = gov.nih.nlm.ncbi.ngs.NGS.openReadCollection(acc.toString());
+            virtualHeader = loadSamHeader();
+            cachedReferences = loadReferences();
+        } catch (final Exception e) {
+            throw new RuntimeException(e);
+        }
+
+        recordRangeInfo = SRAIterator.getRecordsRangeInfo(run);
+        index = new SRAIndex(virtualHeader, recordRangeInfo);
+    }
+
+    @Override
+    public Type type() {
+        return Type.SRA_TYPE;
+    }
+
+    @Override
+    public boolean hasIndex() {
+        return true;
+    }
+
+    @Override
+    public BAMIndex getIndex() {
+        return index;
+    }
+
+    @Override
+    public SAMFileHeader getFileHeader() {
+        return virtualHeader;
+    }
+
+    @Override
+    public CloseableIterator<SAMRecord> getIterator() {
+        return getIterator(new BAMFileSpan(new Chunk(0, recordRangeInfo.getTotalRecordRangeLength())));
+    }
+
+    @Override
+    public CloseableIterator<SAMRecord> getIterator(SAMFileSpan chunks) {
+        if (run == null) {
+            throw new RuntimeException("Cannot create iterator - SRA run is uninitialized");
+        }
+
+        if (virtualHeader == null) {
+            throw new RuntimeException("Cannot create iterator - SAM file header is uninitialized");
+        }
+
+        List<Chunk> chunkList = ((BAMFileSpan) chunks).getChunks();
+
+        final SRAIterator newIterator = new SRAIterator(run, virtualHeader, cachedReferences, recordRangeInfo, chunkList);
+        if (validationStringency != null) {
+            newIterator.setValidationStringency(validationStringency);
+        }
+
+        return newIterator;
+    }
+
+    @Override
+    public SAMFileSpan getFilePointerSpanningReads() {
+        if (recordRangeInfo.getTotalRecordRangeLength() <= 0) {
+            throw new RuntimeException("Cannot create file span - SRA file is empty");
+        }
+
+        return new BAMFileSpan(new Chunk(0, recordRangeInfo.getTotalRecordRangeLength()));
+    }
+
+    @Override
+    public CloseableIterator<SAMRecord> query(QueryInterval[] intervals, boolean contained) { throw new RuntimeException("query is not implemented."); }
```

Comment:
Also, it is rather problematic that this `SamReader` implementation does not support `query()` (despite having an index). This would mean that clients of `htsjdk` would have to special-case `SRAFileReader`s rather than being able to use them transparently via the `SamReader` interface. How difficult would it be to get `query()` working given that you have an index implementation in place?